New select appearances

grzesiek2010 opened the pull request:
#### What has been done to verify that this works as intended?
I tested implemented changes using the form attached below which contains many select one/multiple questions with various appearances.

#### Why is this the best possible solution? Were any other approaches considered?
I implemented new changes according to https://forum.opendatakit.org/t/spec-proposal-harmonize-compact-compact-n-horizontal-horizontal-compact-appearances/15565/13?u=grzesiek2010

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
The implemented changes may affect any select one/multiple widget so that should be carefully tested.
I also added some improvements in handling appearances so other widgets also might be affected - I think testing `all widgets` form would be a good way to confirm there is no regression.

but it might be difficult so as I said above I tested it hardcoding different results.

#### Do we need any specific form for testing your changes? If so, please attach one.
[selectAppearances.zip](https://github.com/opendatakit/collect/files/3182948/selectAppearances.zip)

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
Yes, we will need to document new appearances.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

grzesiek2010 committed:
Support columns-n appearance

grzesiek2010 committed:
Use recyclerView for selects with columns instead of our custom GridWidget

grzesiek2010 committed:
Added tests

grzesiek2010 committed:
Use dividers only if everything is in one column

grzesiek2010 committed:
Fixed bug with media files and autocomplete appearance - wrong media files were used

grzesiek2010 committed:
compact appearance should work like columns-flex

lognaturel commented:
I haven't looked at the details of the `RecyclerView`-based implementation but removing all of that redundant code looks really good.

Keeping all the appearance names straight is rather mind-bending. I do think both `columns` and `columns-flex` are useful and should be different:
* `columns` is what you've implemented here -- some number of columns determined by the screen width
* `columns-flex` is the old `compact` -- it's some number of columns **based on the width of the widest element**

The second one is useful for packing in choices of similar width. In particular, it can work well with image choices.

grzesiek2010 commented:
Ok, I can try to do the same for recyclerView , I mean I can calculate the number of columns based on the width of the widest element.

grzesiek2010 commented:
Ok I was thinking about this and it doesn't make sense to use recyclerview for such compact widgets - we would need to load all elements before anyway to find out how many columns we need, what doesn't make sense of course and recyclerview is for something completely different. I would keep those classes I removed just for compact widgets to use them, in the same way, like before if you want and for all other cases, we can use recyclerview. Do you agree @lognaturel

lognaturel commented:
> we would need to load all elements before anyway to find out how many columns we need

Ah yes, of course. Your approach sounds good and perhaps you can spend a little time thinking about class names to make sure they really capture what each does.

I think we also should pull out all appearances into a single class with static constants at some point. You could consider starting that process here for selects to make it easier to verify that the correct appearance names and deprecated appearance names are supported.

grzesiek2010 commented:
> I think we also should pull out all appearances into a single class with static constants at some point.

I agree. I was thinking about the same some time ago.

grzesiek2010 committed:
Added grid classes back

grzesiek2010 committed:
Created a new class just for appearances

grzesiek2010 committed:
Improved getNumberOfColumns() method

grzesiek2010 committed:
Use compact-n appearance like columns-n that means using recyclerview

grzesiek2010 committed:
Code improvements

grzesiek2010 commented:
@lognaturel I updated the pr and the attached form. Could you take a look? There is only one problem... we agreed above that `columns-flex` should work like `compact` appearance so the easiest way is to use our GridWidgets but using those widgets we can only use images or labels if images are not available- there is no way to show buttons. Do you think it makes sense to add such an option as well? It's just `compact` and `columns-flex` appearances so I'm not sure if we really need it.

lognaturel commented:
Sigh, nothing is ever as simple as it looks, is it?! I do think it’s important to support all variants we agreed on, including the packed in version with buttons.

grzesiek2010 commented:
Ok, so I'll try to use `MediaLayout` in that Grid classes, but after my vacation.

grzesiek2010 committed:
Removed redundand numColumns parameters from GdridWidgets since it's calculated based on items size

grzesiek2010 committed:
Implemented BaseGridWidget class to remove duplicated code

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Support columns-flex appearance using GridWidgets

grzesiek2010 committed:
Fixed GridMultiWidgetTest

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Code improvements

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/WidgetAppearanceUtils.java:
```diff
@@ -0,0 +1,157 @@
+
+    // Get appearance hint and clean it up so it is lower case, without the search function and never null.
+    @NonNull
+    public static String getAppearance(FormEntryPrompt fep) {
```
How about putting this method in `FormEntryPrompt` so it can be called on the prompt itself and giving it a name that makes clear that it transforms the appearance in some way? I'm thinking `getSanitizedAppearanceHint` or something like that. When there's a logical place for behavior, it's best to avoid static methods.

lognaturel replied:
Oh, right, I just remembered `FormEntryPrompt` is in JavaRosa, not Collect. Sigh.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/WidgetAppearanceUtils.java:
```diff
@@ -0,0 +1,157 @@
+        if (appearance.contains(COLUMNS_N) || appearance.contains(COMPACT_N)) {
+            try {
+                String columnsAppearance = appearance.contains(COLUMNS_N) ? COLUMNS_N : COMPACT_N;
+                if (appearance.contains(columnsAppearance)) {
```
I don't think this check is necessary because it's the same as the top-level conditional.

grzesiek2010 replied:
You are right. FIxed.

lognaturel reviewed (commented):
The good news is that I think all of the behavior is there. 

The bad news is that it looks like I flipped `columns` and `columns-flex` at some point and I'm not sure which should be which. I have asked for confirmation at https://forum.opendatakit.org/t/spec-proposal-harmonize-compact-compact-n-horizontal-horizontal-compact-appearances/15565/22?u=ln

One issue: `compact-n` looks like it now has buttons which it should not (it should be identical to `columns-n no-buttons`.

Once you have looked at the `compact-n` issue and made a fix if needed, I think this should go to QA. If we end up needing to swap `columns` and `columns-flex` there will need to be another pass but it can be more targeted.

grzesiek2010 committed:
Compact-n should work without buttons like columns-n no-buttons

grzesiek2010 committed:
Removed redundand if statement and the second parameter in indexOf() method

grzesiek2010 committed:
Fixed bug in Selectmulti widgets - answers were not saved

mmarciniak90 commented:
**Tested with success**

Verified on Android: 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

Verified cases
- new appearance - used the form attached in selectAppearances.zip
- compared widgets on `All widgets` form
- saved answers are visible after opening previously saved form
- removing response

`Autocomplete` is not available for `compact` and `columns-flex`

 appearance |behavior |
|---|---|
|columns |horizontal labels **with** buttons arranged in a number of columns determined by the screen width |
|columns no-buttons |horizontal labels **without** buttons arranged in a number of columns determined by the screen width|
|columns-flex | horizontal labels **with** buttons, packed in to fit horizontal space with minimal padding |
|columns-flex no-buttons |horizontal labels **without** buttons, packed in to fit horizontal space with minimal padding|
|columns-n | horizontal labels **with** buttons arranged in n columns |
|columns-n no-buttons |horizontal labels **without** buttons arranged in n columns |
|compact| don't include radio buttons or checkboxes|
|compact-n| `compact-n` should be identical to `columns-n no-buttons`|

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

lognaturel commented on collect_app/src/test/java/org/odk/collect/android/utilities/FormEntryPromptUtilsTest.java:
```diff
@@ -0,0 +1,68 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class FormEntryPromptUtilsTest {
```
This should be `WidgetAppearanceUtilsTest` now.

grzesiek2010 replied:
Fixed.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/WidgetAppearanceUtils.java:
```diff
@@ -0,0 +1,155 @@
+public class WidgetAppearanceUtils {
+    private static final String EXCEPTION_PARSING_COLUMNS = "Exception parsing columns";
+
+    // Date appearances
```
I wish we'd gone for one or more enums (`DateAppearance`, `SelectAppearance`, etc) here instead of a bunch of constants. Something to maybe come back to at some point. It would also be good to mark the appearances that have been renamed as deprecated (e.g. `compact`, `selfie`).

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/WidgetAppearanceUtils.java:
```diff
@@ -0,0 +1,155 @@
+
+    // Get appearance hint and clean it up so it is lower case, without the search function and never null.
+    @NonNull
+    public static String getAppearance(FormEntryPrompt fep) {
```
A name that makes it clear that there's a transformation would be beneficial. Something like `getSanitizedAppearanceHint`.

grzesiek2010 replied:
Fixed.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/utilities/WidgetAppearanceUtils.java:
```diff
@@ -0,0 +1,155 @@
+        return appearance;
+    }
+
+    public static int getNumberOfColumns(FormEntryPrompt formEntryPrompt, Context context) {
```
A comment would be helpful. Something like "Gets the number of columns that choices should be displayed in. In the case of the columns-n appearance (compact-n for backwards compatibility), that number is determined from the n in the appearance string. For compact-flex, this is determined by the device screen size.

grzesiek2010 replied:
Added.

lognaturel commented:
@grzesiek2010 Thanks, looks good overall! I made a handful of surface comments, mostly about naming so things are easier to understand in the future. Those can be addressed in a future PR if you prefer.

grzesiek2010 committed:
Code improvements

grzesiek2010 commented:
> I made a handful of surface comments, mostly about naming so things are easier to understand in the future. Those can be addressed in a future PR if you prefer.

Thanks, I fixed them apart from just one because I need a bit more time to find a good solution so I'll do that in a separate pr. I also fixed conflicts.
